modules/core/src/main/java/org/apache/ignite/internal/processors/cache/ExchangeContext.java

```diff
@@ -77,6 +95,28 @@ public ExchangeContext(boolean crd, GridDhtPartitionsExchangeFuture fut) {
         evts = new ExchangeDiscoveryEvents(fut);
     }
 
+    /**
+     * @param fut Future.
+     */
+    private boolean localRecoveryNeeded(GridDhtPartitionsExchangeFuture fut) {
+        for (CacheGroupContext grp : fut.sharedContext().cache().cacheGroups()) {
+            if (grp.isLocal())
+                continue;
+
+            GridAffinityAssignmentCache aff = grp.affinity();
+
+            Set<Integer> failedPrimaries = aff.primaryPartitions(fut.exchangeId().eventNode().id(), aff.lastVersion());
+            Set<Integer> locBackups = aff.backupPartitions(fut.sharedContext().localNodeId(), aff.lastVersion());
+
+            for (int part : failedPrimaries) {
+                if (locBackups.contains(part))
+                    return true;
+            }
```

Comment:
This code does not look correct. A node may OWN a partition, but it will not be reported in aff.backupPartitions because the method returns current assignment, not the local partitions state. GridDhtPartitionTopology should be read for this information. A corresponding test need to be added.

Reply:
Copying answer from issue.

Reply:
Copying answer from issue. 
Do you mean we have to serve also moving and renting partitions? 

So, Set<Integer> failedPrimaries = aff.primaryPartitions(fut.exchangeId().eventNode().id(), aff.lastVersion()); is a correct calculation, but
Set<Integer> locBackups = aff.backupPartitions(fut.sharedContext().localNodeId(), aff.lastVersion()); should be replaced with dht.localPartitions() usage?